source/hp/fe_collection.cc

```diff
@@ -27,134 +27,179 @@ namespace hp
+        FiniteElementDomination::Domination domination =
+          FiniteElementDomination::no_requirements;
+        for (const auto &other_fe : fes)
+          if (current_fe != other_fe)
+            domination =
+              domination & finite_elements[current_fe]->compare_for_domination(
                              *finite_elements[other_fe], codim);
 
-          if ((domination ==
-               FiniteElementDomination::other_element_dominates) ||
-              (domination ==
-               FiniteElementDomination::either_element_can_dominate
-               /*covers cases like candidate_fes={Q1,Q1}*/))
-            return current_fe;
-        }
-    // We couldn't find the FE, return invalid_unsigned_int :
+        // If current_fe dominates, return its index.
+        if ((domination == FiniteElementDomination::this_element_dominates) ||
+            (domination == FiniteElementDomination::either_element_can_dominate
+             /*covers cases like {Q2,Q3,Q1,Q1} with fes={2,3}*/))
+          return current_fe;
+      }
+
+    // If we couldn't find the dominating object, return an invalid one.
     return numbers::invalid_unsigned_int;
   }
 
 
 
   template <int dim, int spacedim>
   unsigned int
-  FECollection<dim, spacedim>::find_dominating_fe_in_subset(
+  FECollection<dim, spacedim>::find_dominated_fe(
     const std::set<unsigned int> &fes,
     const unsigned int            codim) const
   {
+    // Validate user inputs.
     Assert(codim <= dim, ExcImpossibleInDim(dim));
-
-    for (const unsigned int fe_index : fes)
+    for (const auto &fe : fes)
       {
-        (void)fe_index;
-        AssertIndexRange(fe_index, finite_elements.size());
+        (void)fe;
+        AssertIndexRange(fe, finite_elements.size());
       }
 
-    // If the set of elements to be dominated contains only a single element X,
-    // then by definition the dominating set contains this single element
-    // (because each element can dominate itself).
-    // There may also be others, in which case we'll check if any of these
-    // elements is able to dominate all others. If one was found, we stop
-    // looking further and return the dominating element.
+    // If the set of elements contains only a single element,
+    // then this very element is considered to be the dominated one.
     if (fes.size() == 1)
       return *fes.begin();
 
-    // loop over all finite elements given in the subset
-    // and check which one dominates the whole subset
+    // There may also be others, in which case we'll check if any of these
+    // elements is dominated by all others. If one was found, we stop
+    // looking further and return the dominated element.
     for (const auto &current_fe : fes)
       {
+        // Check if current_fe is dominated by all other elements in @p fes.
         FiniteElementDomination::Domination domination =
-          FiniteElementDomination::either_element_can_dominate;
```

Comment:
I do not think this was correct in the original code. When you have a look at the conditional below, we assume that there is always a dominating element. This is not always the case, for example for a collection of `{(FE_NothingxFE_Q(1)); (FE_Q(1)xFE_Nothing)}`.

This is the reason why I chose the initial flag to be `no_requirements`. However, this affects the test results of `hp/boundary_matrices_hp`, which I updated in a separate commit. Do you agree with this decision?

Reply:
Note that `hp/boundary_matrices_hp` uses a finite element collection containing `(FE_RaviartThomasNodal(i)xFE_DQG(i)xFE_Nothing)` and `(FE_NothingxFE_NothingxFE_Q(i))` elements with different degrees `i`.